internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"sync"
+	"sync/atomic"
+
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
+	memBuffer         *MemDB
+	flushingMemBuffer *MemDB
```

Comment:
It is a bit subtle to understand how concurrency affects the field, because of the existence of async flushes that run in other goroutines. 
Is it only mutated and accessed in the main thread of memdb since we only allow single-threaded usage, so there is no data race on this field?
Maybe we could make some comments explaning more about this.

Reply:
> Is it only mutated and accessed in the main thread of memdb since we only allow single-threaded usage, so there is no data race on this field?

Yes, the usage is same as the memdb, no concurrently read & write.